<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [2]:
import pandas as pd
import numpy as np
import psycopg2
import plotly.express as px
import seaborn as sns
import matplotlib.pyplot as plt

In [25]:

# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

In [26]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [18]:
# текст запроса
query_3_1 = '''select count(*)
           from vacancies
'''

In [19]:
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_22832\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [20]:
# текст запроса
query_3_2 = '''select count(*)
           from EMPLOYERS
'''

In [21]:
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_22832\3588252056.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [22]:
# текст запроса
query_3_3 = '''select count(*)
           from areas
'''

In [23]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_22832\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [24]:
# текст запроса
query_3_4 = '''select count(*)
           from industries
'''

In [25]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_22832\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

Вывод по предварительному анализу:
1) Работодателей почти в 2 раза меньше чем вакансий, это значит, что один работодатель предлагает в среднем 2-3 вакансии

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [26]:
# текст запроса
query_4_1 = '''select count(*),
                     a.name
                from areas a
                left join vacancies v on v.area_id = a.id
                group by a.name
                order by 1 desc
                limit 5
'''

In [27]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_22832\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


count             name
0   5333           Москва
1   2851  Санкт-Петербург
2   2112            Минск
3   2006      Новосибирск
4   1892           Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [28]:
# текст запроса
query_4_2 = '''
select 
    count(*)
from vacancies v
where salary_from is not null
or salary_to is not null'''

In [29]:
# результат запроса
df = pd.read_sql_query(query_4_2,connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_22832\949731072.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2,connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [30]:
# текст запроса
query_4_3 = '''
select 
    ROUND(avg(v.salary_from))
from vacancies v

union

select 
    ROUND(avg(v.salary_to))
from vacancies v'''

In [31]:
# результат запроса
df = pd.read_sql_query(query_4_3,connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_22832\3309279565.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3,connection)


round
0   71065.0
1  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [32]:
# текст запроса
query_4_4 = '''
select distinct
    schedule,
    employment,
    count(*)
from vacancies
group by schedule,employment
order by 3 desc
limit 2'''


In [33]:
# результат запроса
df = pd.read_sql_query(query_4_4,connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_22832\2166644490.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4,connection)


schedule        employment  count
0       Полный день  Полная занятость  35367
1  Удаленная работа  Полная занятость   7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [34]:
# текст запроса
query_4_5 = '''
select
    experience,
    count (*)
from vacancies
where experience is not null
group by experience
order by 2'''

In [35]:
# результат запроса
df = pd.read_sql_query(query_4_5,connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_22832\1722615915.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5,connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

Вывод:
1) Мы узнали распределиние количество вакансиий в городах, и поняли что больше всего вакансий в столице и культурной столице России, а так же имеет место влияние количество населения в городе.
2) Количество вакансий, у которых заполнено хотя бы 1 из полей вилки заработной платы, примерно половино всех вакансий
3) Средняя вилка заработной платы состовляет 71065 рублей 110537 рублей
4) Определили, что популярнее всего вакансии с комбинацией "Полный день-Полная занятость"
5) Наглядно видно, что в поле опыт работы меньше всего ищут работу соискатели с опытом работы "более 6 лет", это связано с тем, что с опытом работы более лет уже соискатель в основном обрастает связями и не пользуется услугами Hh

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [36]:
# текст запроса
query_5_1 = '''
select 
    e.name,
    count(v.id)
from vacancies v
left join employers e on v.employer_id = e.id
group by e.name
order by 2 desc
limit 5'''

In [37]:
# результат запроса
df = pd.read_sql_query(query_5_1,connection)
display(df['name'][0])
display(df['name'][4])

C:\Users\User\AppData\Local\Temp\ipykernel_22832\3399669460.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1,connection)


'Яндекс'

'Газпром нефть'

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [38]:
# текст запроса
query_5_2 = '''
select a.name,
       count(distinct e.id) cnt_emp,
       count(distinct v.id) cnt_vac
from areas a
     left join employers e on a.id = e.area
     left join vacancies v on v.area_id = a.id
group by a.id
order by cnt_emp desc
'''

In [39]:
# результат запроса
df = pd.read_sql_query(query_5_2,connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_22832\3039681661.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2,connection)


name  cnt_emp  cnt_vac
0                                  Москва     5864     5333
1                         Санкт-Петербург     2217     2851
2                                   Минск     1115     2112
3                                  Алматы      721     1892
4                            Екатеринбург      609     1698
...                                   ...      ...      ...
1357                              Янино-1        0        0
1358                          Адыге-Хабль        0        1
1359                          Эркен-Шахар        0        0
1360  Индустриальный (Краснодарский край)        0        0
1361          Сабетта (Ямало-Ненецкий АО)        0        0

[1362 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [40]:
# текст запроса
query_5_3 = '''
select 
    e.name,
    count(distinct a.name) "Количество регионов"
from vacancies v
left join employers e on v.employer_id = e.id
left join areas a on a.id = v.area_id
group by e.name
order by 2 desc'''

In [41]:
# результат запроса
df_1 = pd.read_sql_query(query_5_3,connection)
df_1

C:\Users\User\AppData\Local\Temp\ipykernel_22832\1916418514.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_1 = pd.read_sql_query(query_5_3,connection)


name  Количество регионов
0                      Яндекс                  181
1                  Ростелеком                  152
2                  Спецремонт                  116
3      Поляков Денис Иванович                   88
4                   ООО ЕФИН                    71
...                       ...                  ...
14761                  UniSol                    1
14762            UNISTORY LLC                    1
14763                   UNIT6                    1
14764     United Distribution                    1
14765              UNITEL LLC                    1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [42]:
# текст запроса
query_5_4 = '''select 
    count(*) 
from industries i
full join employers_industries ei on i.id = ei.industry_id
full join employers e on ei.employer_id = e.id
where i.name is null'''

In [43]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_22832\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [44]:
# текст запроса
query_5_5 = '''
select 
    e.name,
    count(i.id)
from industries i
full join employers_industries ei on i.id = ei.industry_id
full join employers e on ei.employer_id = e.id
group by e.name
having count(i.id) = 4
order by 1
offset 2 limit 1 '''

In [45]:
# результат запроса
df = pd.read_sql_query(query_5_5,connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_22832\2182950375.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5,connection)


name  count
0  2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [46]:
# текст запроса
query_5_6 = '''
select 
    count(*)
from industries i
full join employers_industries ei on i.id = ei.industry_id
full join employers e on ei.employer_id = e.id
where i.name = 'Разработка программного обеспечения'
'''

In [69]:
# результат запроса
df = pd.read_sql_query(query_5_6,connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_22832\2938330085.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6,connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [ ]:
# код для получения списка городов-милионников
import urllib.parse as req
import requests
from bs4 import BeautifulSoup

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)
response = requests.get(url) # Выполняем GET-запрос, содержимое ответа присваивается переменной response
tables = pd.read_html(response)

In [72]:
tables = pd.read_html('https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8')
df = tables[1]
cities = list(df['Город'])
cities[-3] = 'Воронеж'
cities = tuple(cities)
print(cities)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')


In [73]:
# текст запроса
query_5_7 =f'''
select 
    a.name as name,
    count(v.id)
from vacancies v
left join employers e on v.employer_id = e.id
left join areas a on a.id = v.area_id
where e.name = 'Яндекс'
and a.name in {cities}
group by a.name

union all

select 
    'total' as name,
    count(*)
from vacancies v
left join employers e on v.employer_id = e.id
left join areas a on a.id = v.area_id
where e.name = 'Яндекс'
and a.name in {cities}
order by 2
'''

In [74]:
# результат запроса
df = pd.read_sql_query(query_5_7,connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_12168\2337172344.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7,connection)


name  count
0              Омск     21
1         Челябинск     22
2        Красноярск     23
3         Волгоград     24
4             Пермь     25
5            Казань     25
6    Ростов-на-Дону     25
7               Уфа     26
8            Самара     26
9         Краснодар     30
10          Воронеж     32
11      Новосибирск     35
12  Нижний Новгород     36
13     Екатеринбург     39
14  Санкт-Петербург     42
15           Москва     54
16            total    485

***

Вывод:
1) Определили, что больше всего вакансий предлагает "Яндекс"-это означает, что в "Яндексе" требуется большое количество специалистов.
2) В Москве и Санкт-Петербурге больше всего работодателей и вакансий. 

In [ ]:
fig = px.treemap(
    data_frame=df_1[:100], #датафрейм
    path=['name'], #кат. признак, для которого строится график(на углу прямоугольничка инфа)
    values='Количество регионов', #параметр, который сравнивается(т.е. та инфа которая будет отображаться
    #на прямоугольничке)
    height=500, #высота
    width=700, #ширина
    title='Daily Recovered Cases by Country' #заголовок
)

fig.show()

На графике видны отношения работодателей к регионам(топ 100), отлично видно, что "Яндекс" и "Ростелеком" занимают лидирующие места и распростроняют свои вакансии по стране

Это также значит, что у Яндекса большой штат работников и Яндекс многогранен в своих разработках. 

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [ ]:
# текст запроса
query_6_1 = '''
select
    count(*)
from vacancies v
where lower(v.name) like '%data%'
or lower(v.name) like '%данн%'
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_1,connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_18484\1520672217.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
# текст запроса
query_6_2 ='''
select
    count(*)
from vacancies v
where (lower(v.name) like '%data scientist%'
or lower(v.name) like '%data science%'
or lower(v.name) like '%исследователь данных%'
or lower(v.name) like '%machine learning%'
or lower(v.name) like '%машинн%обучен%'
or lower(v.name) not like '%html%'
and lower(v.name) like '%ml%')
and
(lower(v.name) like '%junior%'
or v.experience = 'Нет опыта'
or v.employment = 'Стажировка')'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_2,connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_18484\1041952921.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# текст запроса
query_6_3 = '''
select
    count(*)
from vacancies v
where (lower(v.name) like '%data%scientist%'
or lower(v.name) like '%data%science%'
or lower(v.name) like '%исследователь%данных%'
or lower(v.name) like '%machine%learning%'
or lower(v.name) like '%машинн%обучен%'
or lower(v.name) not like '%html%'
and v.name like '%ML%')
and
(lower(v.key_skills) like '%sql%'
or lower(v.key_skills) like '%postgres%')
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_3,connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_18484\3683206453.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# текст запроса
query_6_4 = '''
select
    count(*)
from vacancies v
where (lower(v.name) like '%data%scientist%'
or lower(v.name) like '%data%science%'
or lower(v.name) like '%исследователь%данных%'
or lower(v.name) like '%machine%learning%'
or lower(v.name) like '%машинн%обучен%'
or lower(v.name) not like '%html%'
and v.name like '%ML%')
and
(lower(v.key_skills) like '%python%'
)
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_4,connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_18484\3924223682.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
# текст запроса
query_6_5 = '''
select
    round(avg(length(key_skills)
             -length(replace(key_skills, CHR(9), ''))+1), 2)
from public.vacancies
where 
(lower(name) like '%data%scientist%'
or lower(name) like '%data%science%'
or lower(name) like '%исследователь%данных%'
or lower(name) like '%machine%learning%'
or lower(name) like '%машинн%обучен%'
or lower(name) not like '%html%'
and name like '%ML%')'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_5,connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_18484\965758319.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [35]:
# текст запроса
query_6_6 = '''
select
    experience,
    round(avg(coalesce((salary_from+salary_to)/2,salary_from,salary_to)))
from vacancies
where 
(lower(name) like '%data%scientist%'
or lower(name) like '%data%science%'
or lower(name) like '%исследователь%данных%'
or lower(name) like '%machine%learning%'
or lower(name) like '%машинн%обучен%'
or lower(name) not like '%html%'
and name like '%ML%')
group by experience'''

In [36]:
# результат запроса
df = pd.read_sql_query(query_6_6,connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_12168\943827912.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6,connection)


experience     round
0         Более 6 лет       NaN
1           Нет опыта   74643.0
2  От 1 года до 3 лет  139675.0
3       От 3 до 6 лет  243115.0

***

Согласно таблице можем сделать вывод, что зарплата у Data Scientist растет в геометрической прогрессии :) с ростом опыта работы.
Также заметим, что зарплата не указывается при опыте работы более 6 лет. Это возможно из-за того, что Data Scientist с опытом работы более 6 лет имеет большой background и возникает сложность при определении заработной платы.

# Общий вывод по проекту

Общий вывод по проекту можно сделать следующий:

Лидирующие места на бирже труда занимают корпорации занимающиеся в IT сфере, значит данные профессии востребованнее всего. К тому же можно сделать прогноз, что цифры, отражающие потребность в IT-шниках, будут расти.

Общая потребность в людях специализирующихся в Data Science высокая и эта потребность будет расти. Навыки Data Scientist-ов будут вытеснять такие специальности, как бэкенд разработчик и фронтэнд разработчик(с помощью нейроннных сетей). Нужно подчеркнуть, что каждому малому, среднему и крупному бизнесу нужна команда Data Scientist-ов, так как они анализируют рынок и помогают делать рациональные выводы по развитию бизнеса.